File: App/Resources/AlgorithmDetails/threesmoothcombsortiterative/threesmoothcombsortiterative.go
package main

import (
	"fmt"
	"math"
)

func sort(arr []int) []int {
	n := len(arr)
	if n <= 1 {
		return arr
	}
	pow2 := int(math.Log(float64(n-1)) / math.Log(2))
	for k := pow2; k >= 0; k-- {
		pow3 := int((math.Log(float64(n)) - float64(k)*math.Log(2)) / math.Log(3))
		for j := pow3; j >= 0; j-- {
			gap := int(math.Pow(2, float64(k)) * math.Pow(3, float64(j)))
			for i := 0; i+gap < n; i++ {
				if arr[i] > arr[i+gap] {
					arr[i], arr[i+gap] = arr[i+gap], arr[i]
				}
			}
		}
	}
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}

File: App/Resources/AlgorithmDetails/threesmoothcombsortiterative/threesmoothcombsortiterative.java
import java.util.Arrays;

public class threesmoothcombsortiterative {
  public static void sort(int[] arr) {
    int n = arr.length;
    if (n <= 1) {
      return;
    }
    int pow2 = (int) (Math.log(n - 1) / Math.log(2));
    for (int k = pow2; k >= 0; k--) {
      int pow3 = (int) ((Math.log(n) - k * Math.log(2)) / Math.log(3));
      for (int j = pow3; j >= 0; j--) {
        int gap = (int) (Math.pow(2, k) * Math.pow(3, j));
        for (int i = 0; i + gap < n; i++) {
          if (arr[i] > arr[i + gap]) {
            int t = arr[i];
            arr[i] = arr[i + gap];
            arr[i + gap] = t;
          }
        }
      }
    }
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
